File: union_find/weighted_qu.py
import unittest

class QuickUnionUF:

    def __init__(self, n):
      self.n = n
      self.array = list(range(n))
      self.size = { key: 0 for key in self.array }

    def connected(self, x, y):
      return self.__root(x) == self.__root(y)

    def union(self, x, y):
      i, j = self.__root(x), self.__root(y)
      if i == j:
        return
      elif self.size[i] < self.size[j]:
        self.array[i] = j
        self.size[j] += self.size[i]
      else:
        self.array[j] = i
        self.size[i] += self.size[j]

    def __root(self, i):
      while i != self.array[i]:
        i = self.array[i]
      return i

class QuickUnionUFTest(unittest.TestCase):
  
    def setUp(self):
        self.obj = QuickUnionUF(10)

    def test_initialize(self):
        expected = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
        self.assertEqual(self.obj.array, expected)

    def test_connected(self):
        self.assertFalse(self.obj.connected(2, 3))

    def test_union(self):
        self.obj.union(2, 3)
        self.assertTrue(self.obj.connected(2, 3))

if __name__ == '__main__':
    unittest.main()

File: union_find/weighted_qu.rb
require 'test/unit'

class QuickUnionUF
 
  attr_reader :array

  def initialize n
    @n = n-1
    @array = (0..@n).to_a
    @size = {}
    @array.each do |k|
      @size[k] = 0
    end
  end

  def connected x, y
    root(x) == root(y)
  end

  def union x, y
    i, j = root(x), root(y)
    return if i == j
    if @size[i] < @size[j]
      @array[i] = j
      @size[j] += @size[i]
    else
      @array[j] = i
      @size[i] += @size[j]
    end
  end

  private

  def root (i)
    i = @array[i] until i == @array[i]
    return i
  end

end


class QuickUnionUFTest < Test::Unit::TestCase
  
  def setup
    @obj = QuickUnionUF.new(10)
  end

  def test_initialize
    expected = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    assert_equal(expected, @obj.array)
  end

  def test_connected
    assert_equal(false, @obj.connected(2, 3))
  end

  def test_union
    @obj.union(2, 3)
    assert_equal(true, @obj.connected(2, 3))
  end

end